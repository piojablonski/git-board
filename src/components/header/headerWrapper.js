import styled, { css } from 'styled-components'
import * as vars from '../../styled/variables'

export const HeaderWrapper = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: flex-end;
  position: fixed;
  top: 0;
  width: 100%;
  background-color: ${vars.gray3};
  height: ${vars.headerHeight};
  z-index: 1;

  > * {
    margin-right: 0.5rem;
  }

  .menu-icon {
    font-size: 1.5em;
    position: absolute;
    left: 0.75rem;
    line-height: ${vars.headerHeight};
    ${vars.media.tablet`
      display: ${props => props.isSidebarOpened ? 'none' : 'block'};
    `}
  }
`