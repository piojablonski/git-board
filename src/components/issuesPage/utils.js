export const mapDataToOption = (value, name) => res => res.data.map(l => ({ value: String(l[value]), name: String(l[name]) })) || []
export const sanitizeSelectedFilters = selectedFilters => {
  const res = Object.keys(selectedFilters).reduce((acc, key) => {
    const value = selectedFilters[key]
    if (Array.isArray(value)) {
      acc[key] = value.join()
    } else if (value && value !== '') {
      acc[key] = value
    }
    return acc
  }, {})
  return res
}

export const readPaginationData = response => {
  const link = response.headers.link
  if (!link) {
    return 1
  }
  const linksArr = link.split(',')
  const lastPage = linksArr.reduce((acc, link) => {
    const page = link.match(/page=(\d+).*rel="last"$/)
    if (page && page.length > 0) {
      return page[1]
    }
    return acc
  }, undefined)
  return lastPage
}