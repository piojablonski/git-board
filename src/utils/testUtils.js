
export const matchSnapshot = (component) => {
  expect(component).toMatchSnapshot()
}
