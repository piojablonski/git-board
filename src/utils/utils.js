import { setKind, redirect } from 'redux-first-router'

export const navigate = (type, payload, query) => {
  const result = setKind(redirect({ type, payload, meta: { query } }), 'push')
  return result
}
