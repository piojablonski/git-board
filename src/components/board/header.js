import React from 'react'
import { connect } from 'react-redux'
import { redirect } from 'redux-first-router'
import { issuesActions } from '../../reducers/issues.reducer'
import { HeaderWrapper } from './styled'
import { Button, Select, Icon } from 'antd'
import { navigate } from '../../utils/utils'
const { Option } = Select

export const HeaderComponent = ({ pagination, pageSizeOptions, filterChangeAndRedirect, selectedFilters }) => {
  const executeChange = (key) => (value) => {
    filterChangeAndRedirect(selectedFilters, value, 'per_page', { page: 1 })
  }
  const executeChangePage = ({ target: { value } }) => {
    filterChangeAndRedirect(selectedFilters, value, 'page')
  }
  return (
    <HeaderWrapper>
      <Button type="primary" disabled={!pagination.hasPrevious} value={pagination.previousPage} onClick={executeChangePage}>
        <Icon type="left" />Previous
      </Button>
      <span>{`${pagination.page} / ${pagination.lastPage}`}</span>
      <Button type="primary" disabled={!pagination.hasNext} value={pagination.nextPage} onClick={executeChangePage}>
        Next<Icon type="right" />
      </Button>
      <Select
        value={pagination.perPage}
        onChange={executeChange('per_page')} >
        {pageSizeOptions.map(p => <Option key={p} value={p}>{p}</Option>)}
      </Select>
    </HeaderWrapper>
  )
}

const mapStateToProps = (state) => ({
  pageSizeOptions: state.issues.options.static.perPage,
  pagination: paginationInfoSelector(state.issues),
  selectedFilters: state.issues.selectedFilters

})
const mapDispatchToProps = (dispatch) => ({
  filterChangeAndRedirect: (query, value, filterKey, extraValues = {}) => {
    dispatch(navigate('ISSUES', { ...query, ...extraValues, [filterKey]: value }))
  }
})

const paginationInfoSelector = issuesState => {
  const { lastPage, selectedFilters: { page, per_page } } = issuesState
  const nLastPage = Number(lastPage)
  const nPage = Number(page)
  const res = ({
    hasPrevious: nPage > 1,
    previousPage: String(nPage - 1),
    currentPage: page,
    lastPage,
    page,
    hasNext: nPage < nLastPage,
    nextPage: String(nPage + 1),
    perPage: per_page
  })
  return res
}

export const Header = connect(mapStateToProps, mapDispatchToProps)(HeaderComponent)
