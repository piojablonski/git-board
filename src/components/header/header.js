import React from 'react'
import { connect } from 'react-redux'
import { issuesActions } from '../../reducers/issues.reducer'
import { HeaderWrapper } from './headerWrapper'
import { Button, Select, Icon } from 'antd'
import { navigate } from '../../utils/utils'
const { Option } = Select

export const HeaderComponent = ({ pagination, ...props }) => {
  const executeChange = (key) => (value) => {
    props.filterChangeAndRedirect(props.selectedFilters, value, 'per_page', { page: 1 })
  }
  const executeChangePage = ({ target: { value } }) => {
    props.filterChangeAndRedirect(props.selectedFilters, value, 'page')
  }

  return (
    <HeaderWrapper isSidebarOpened={props.isSidebarOpened}>
      <Icon type='menu-unfold' onClick={props.toggleSidebar} className='menu-icon' />
      <Button type="default" disabled={!pagination.hasPrevious} value={pagination.previousPage} onClick={executeChangePage}>
        <Icon type="left" />Prev
      </Button>
      <span>{`${pagination.page} / ${pagination.lastPage || ''}`}</span>
      <Button type="default" disabled={!pagination.hasNext} value={pagination.nextPage} onClick={executeChangePage}>
        Next<Icon type="right" />
      </Button>
      <Select
        value={pagination.perPage}
        onChange={executeChange('per_page')} >
        {props.pageSizeOptions.map(p => <Option key={p} value={p}>{p}</Option>)}
      </Select>
    </HeaderWrapper>
  )
}

const mapStateToProps = (state) => ({
  pageSizeOptions: state.issues.options.static.perPage,
  pagination: paginationInfoSelector(state.issues),
  selectedFilters: state.issues.selectedFilters,
  isSidebarOpened: state.issues.isSidebarOpened

})
const mapDispatchToProps = (dispatch) => ({
  filterChangeAndRedirect: (query, value, filterKey, extraValues = {}) => {
    dispatch(navigate('ISSUES', { ...query, ...extraValues, [filterKey]: value }))
  },
  toggleSidebar: () => { dispatch(issuesActions.toggleSidebar()) }
})

const paginationInfoSelector = issuesState => {
  const { lastPage, selectedFilters: { page, per_page } } = issuesState
  const nLastPage = Number(lastPage)
  const nPage = Number(page)
  const res = ({
    hasPrevious: nPage > 1,
    previousPage: String(nPage - 1),
    lastPage,
    page,
    hasNext: nPage < nLastPage,
    nextPage: String(nPage + 1),
    perPage: per_page
  })
  return res
}

export const Header = connect(mapStateToProps, mapDispatchToProps)(HeaderComponent)
