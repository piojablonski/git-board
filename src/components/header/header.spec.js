import React from 'react'
// import 'jest-styled-components'
import { HeaderComponent } from './header'
import { mount, shallow } from 'enzyme'
import { matchSnapshot } from '../../utils/testUtils'
import { initialIssuesState } from '../../reducers/issues.reducer'
import { Button } from 'antd'

const defaultProps = () => ({
  pageSizeOptions: ['10', '20', '30'],
  pagination: {
    hasPrevious: true,
    previousPage: '1',
    page: '2',
    nextPage: '3',
    lastPage: '4',
    hasNext: true,
    perPage: '20'
  },
  selectedFilters: initialIssuesState.selectedFilters,
  isSidebarOpened: false,
  filterChangeAndRedirect: jest.fn(),
  toggleSidebar: jest.fn(),
  routeType: 'ISSUE'

})

const prevButton = c => c.find(Button).at(0)
const nextButton = c => c.find(Button).at(1)
const menuButton = c => c.find({ type: 'menu-unfold' })

describe('HeaderComponent', () => {
  it('should have both buttons enabled', () => {
    const comp = <HeaderComponent {...defaultProps() } />
    const tree = shallow(comp)
    matchSnapshot(tree)
    expect(prevButton(tree).props().disabled).toBe(false)
    expect(nextButton(tree).props().disabled).toBe(false)
  })

  it('should have previous button disabled', () => {
    const props = {
      ...defaultProps(),
      pagination: {
        hasPrevious: false,
        previousPage: undefined,
        page: '1',
        nextPage: '2',
        lastPage: '3',
        hasNext: true,
        perPage: '20'
      }
    }

    const comp = <HeaderComponent {...props} />
    const tree = shallow(comp)
    matchSnapshot(tree)

    expect(prevButton(tree).props().disabled).toBe(true)
    expect(nextButton(tree).props().disabled).toBe(false)
  })

  it('should have next button disabled', () => {
    const props = {
      ...defaultProps(),
      pagination: {
        hasPrevious: true,
        previousPage: '1',
        page: '2',
        nextPage: undefined,
        lastPage: '2',
        hasNext: false,
        perPage: '20'
      }
    }

    const comp = <HeaderComponent {...props} />
    const tree = shallow(comp)
    matchSnapshot(tree)

    expect(prevButton(tree).props().disabled).toBe(false)
    expect(nextButton(tree).props().disabled).toBe(true)
  })

  it('should have prev button that calls filterChangeAndRedirect on click', () => {
    const props = defaultProps()
    const comp = mount(<HeaderComponent {...props} />)
    prevButton(comp).simulate('click')
    expect(props.filterChangeAndRedirect).toHaveBeenCalled()
  })

  it('should have menu button that calls toggleSidebar on click', () => {
    const props = defaultProps()
    const comp = mount(<HeaderComponent {...props} />)
    menuButton(comp).simulate('click')
    expect(props.toggleSidebar).toHaveBeenCalled()
  })
})
