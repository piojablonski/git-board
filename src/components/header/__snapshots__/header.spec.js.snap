// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderComponent should have both buttons enabled 1`] = `
<styled.div
  isSidebarOpened={false}
>
  <Icon
    className="menu-icon"
    onClick={[Function]}
    type="menu-unfold"
  />
  <Button
    disabled={false}
    ghost={false}
    loading={false}
    onClick={[Function]}
    prefixCls="ant-btn"
    type="default"
    value="1"
  >
    <Icon
      type="left"
    />
    Prev
  </Button>
  <span>
    2 / 4
  </span>
  <Button
    disabled={false}
    ghost={false}
    loading={false}
    onClick={[Function]}
    prefixCls="ant-btn"
    type="default"
    value="3"
  >
    Next
    <Icon
      type="right"
    />
  </Button>
  <Select
    choiceTransitionName="zoom"
    onChange={[Function]}
    prefixCls="ant-select"
    showSearch={false}
    transitionName="slide-up"
    value="20"
  >
    <Option
      key="10"
      value="10"
    >
      10
    </Option>
    <Option
      key="20"
      value="20"
    >
      20
    </Option>
    <Option
      key="30"
      value="30"
    >
      30
    </Option>
  </Select>
</styled.div>
`;

exports[`HeaderComponent should have next button disabled 1`] = `
<styled.div
  isSidebarOpened={false}
>
  <Icon
    className="menu-icon"
    onClick={[Function]}
    type="menu-unfold"
  />
  <Button
    disabled={false}
    ghost={false}
    loading={false}
    onClick={[Function]}
    prefixCls="ant-btn"
    type="default"
    value="1"
  >
    <Icon
      type="left"
    />
    Prev
  </Button>
  <span>
    2 / 2
  </span>
  <Button
    disabled={true}
    ghost={false}
    loading={false}
    onClick={[Function]}
    prefixCls="ant-btn"
    type="default"
  >
    Next
    <Icon
      type="right"
    />
  </Button>
  <Select
    choiceTransitionName="zoom"
    onChange={[Function]}
    prefixCls="ant-select"
    showSearch={false}
    transitionName="slide-up"
    value="20"
  >
    <Option
      key="10"
      value="10"
    >
      10
    </Option>
    <Option
      key="20"
      value="20"
    >
      20
    </Option>
    <Option
      key="30"
      value="30"
    >
      30
    </Option>
  </Select>
</styled.div>
`;

exports[`HeaderComponent should have previous button disabled 1`] = `
<styled.div
  isSidebarOpened={false}
>
  <Icon
    className="menu-icon"
    onClick={[Function]}
    type="menu-unfold"
  />
  <Button
    disabled={true}
    ghost={false}
    loading={false}
    onClick={[Function]}
    prefixCls="ant-btn"
    type="default"
  >
    <Icon
      type="left"
    />
    Prev
  </Button>
  <span>
    1 / 3
  </span>
  <Button
    disabled={false}
    ghost={false}
    loading={false}
    onClick={[Function]}
    prefixCls="ant-btn"
    type="default"
    value="2"
  >
    Next
    <Icon
      type="right"
    />
  </Button>
  <Select
    choiceTransitionName="zoom"
    onChange={[Function]}
    prefixCls="ant-select"
    showSearch={false}
    transitionName="slide-up"
    value="20"
  >
    <Option
      key="10"
      value="10"
    >
      10
    </Option>
    <Option
      key="20"
      value="20"
    >
      20
    </Option>
    <Option
      key="30"
      value="30"
    >
      30
    </Option>
  </Select>
</styled.div>
`;
